File: exam-slammr-webapp/src/test/java/uk/co/claritysoftware/exam/slammr/webapp/web/controller/QuestionControllerTest.java
package uk.co.claritysoftware.exam.slammr.webapp.web.controller;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.mockito.Mockito.mock;
import static uk.co.claritysoftware.exam.slammr.webapp.testsupport.service.model.question.QuestionTestDataFactory.aSimpleQuestionAboutTriangles;
import static uk.co.claritysoftware.exam.slammr.webapp.testsupport.web.model.question.CreateQuestionTestDataFactory.aSimpleCreateQuestionAboutTriangles;
import static uk.co.claritysoftware.exam.slammr.webapp.web.model.question.AnswerOption.generateEmptyAnswerOption;
import static uk.co.claritysoftware.exam.slammr.webapp.web.model.question.CreateQuestion.generateEmptyCreateQuestion;
import static uk.co.claritysoftware.exam.slammr.webapp.web.model.question.FurtherReading.generateEmptyFurtherReading;

import java.security.Principal;
import java.util.ArrayList;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.ui.ExtendedModelMap;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.servlet.ModelAndView;
import uk.co.claritysoftware.exam.slammr.webapp.service.QuestionService;
import uk.co.claritysoftware.exam.slammr.webapp.service.model.question.Question;
import uk.co.claritysoftware.exam.slammr.webapp.service.model.question.QuestionStatus;
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.AnswerOption;
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.CreateQuestion;
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.FurtherReading;

/**
 * Unit test class for {@link QuestionController}
 */
@RunWith(MockitoJUnitRunner.class)
public class QuestionControllerTest {

	@Mock
	private QuestionService questionService;

	@InjectMocks
	private QuestionController questionController;

	@Test
	public void shouldGetQuestionPage() {
		// Given
		CreateQuestion emptyQuestionForm = generateEmptyCreateQuestion();
		Model expectedModel = new ExtendedModelMap();
		expectedModel.addAttribute("form", emptyQuestionForm);
		expectedModel.addAttribute("showBindErrors", false);

		// When
		ModelAndView modelAndView = questionController.getQuestionPage();

		// Then
		assertThat(modelAndView)
				.as("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(asList(
						"question/create",
						expectedModel
				));
	}

	@Test
	public void shouldNotCreateNewQuestionGivenAddTagAction() {
		// Given
		BindingResult bindingResult = mock(BindingResult.class);
		Principal principal = mock(Principal.class);
		CreateQuestion createQuestionForm = createQuestion()
				.tags(new ArrayList<>(singletonList("tag1")))
				.action(CreateQuestion.Action.addTag)
				.build();

		CreateQuestion returnedCreateQuestionForm = createQuestion()
				.tags(asList("tag1", ""))
				.action(CreateQuestion.Action.addTag)
				.build();
		Model expectedModel = new ExtendedModelMap();
		expectedModel.addAttribute("form", returnedCreateQuestionForm);
		expectedModel.addAttribute("showBindErrors", false);

		// When
		ModelAndView modelAndView = questionController.createNewQuestion(createQuestionForm, bindingResult, principal);

		// Then
		assertThat(modelAndView)
				.as("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(asList(
						"question/create",
						expectedModel
				));
	}

	@Test
	public void shouldNotCreateNewQuestionGivenAddCertificationAction() {
		// Given
		BindingResult bindingResult = mock(BindingResult.class);
		Principal principal = mock(Principal.class);
		CreateQuestion createQuestionForm = createQuestion()
				.certifications(new ArrayList<>(singletonList("certification1")))
				.action(CreateQuestion.Action.addCertification)
				.build();

		CreateQuestion returnedCreateQuestionForm = createQuestion()
				.certifications(asList("certification1", ""))
				.action(CreateQuestion.Action.addCertification)
				.build();
		Model expectedModel = new ExtendedModelMap();
		expectedModel.addAttribute("form", returnedCreateQuestionForm);
		expectedModel.addAttribute("showBindErrors", false);

		// When
		ModelAndView modelAndView = questionController.createNewQuestion(createQuestionForm, bindingResult, principal);

		// Then
		assertThat(modelAndView)
				.as("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(asList(
						"question/create",
						expectedModel
				));
	}

	@Test
	public void shouldNotCreateNewQuestionGivenAddFurtherReadingAction() {
		// Given
		BindingResult bindingResult = mock(BindingResult.class);
		Principal principal = mock(Principal.class);
		CreateQuestion createQuestionForm = createQuestion()
				.furtherReadings(new ArrayList<>(singletonList(FurtherReading.builder()
						.description("Some reference")
						.referenceLocation("http://some.location")
						.build())))
				.action(CreateQuestion.Action.addFurtherReading)
				.build();

		CreateQuestion returnedCreateQuestionForm = createQuestion()
				.furtherReadings(asList(FurtherReading.builder()
								.description("Some reference")
								.referenceLocation("http://some.location")
								.build(),
						FurtherReading.builder().build()))
				.action(CreateQuestion.Action.addFurtherReading)
				.build();
		Model expectedModel = new ExtendedModelMap();
		expectedModel.addAttribute("form", returnedCreateQuestionForm);
		expectedModel.addAttribute("showBindErrors", false);

		// When
		ModelAndView modelAndView = questionController.createNewQuestion(createQuestionForm, bindingResult, principal);

		// Then
		assertThat(modelAndView)
				.as("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(asList(
						"question/create",
						expectedModel
				));
	}

	@Test
	public void shouldNotCreateNewQuestionGivenAddAnswerAction() {
		// Given
		BindingResult bindingResult = mock(BindingResult.class);
		Principal principal = mock(Principal.class);
		CreateQuestion createQuestionForm = createQuestion()
				.answerOptions(new ArrayList<>(singletonList(AnswerOption.builder()
						.answer("Some answser")
						.correct(true)
						.build())))
				.action(CreateQuestion.Action.addAnswer)
				.build();

		CreateQuestion returnedCreateQuestionForm = createQuestion()
				.answerOptions(asList(AnswerOption.builder()
								.answer("Some answser")
								.correct(true)
								.build(),
						AnswerOption.builder().build()))
				.action(CreateQuestion.Action.addAnswer)
				.build();
		Model expectedModel = new ExtendedModelMap();
		expectedModel.addAttribute("form", returnedCreateQuestionForm);
		expectedModel.addAttribute("showBindErrors", false);

		// When
		ModelAndView modelAndView = questionController.createNewQuestion(createQuestionForm, bindingResult, principal);

		// Then
		assertThat(modelAndView)
				.as("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(asList(
						"question/create",
						expectedModel
				));
	}

	@Test
	public void shouldNotCreateNewQuestionGivenSaveActionAndBindingErrors() {
		// Given
		BindingResult bindingResult = mock(BindingResult.class);
		Principal principal = mock(Principal.class);
		CreateQuestion createQuestionForm = aSimpleCreateQuestionAboutTriangles()
				.summary(null)
				.action(CreateQuestion.Action.save)
				.build();

		given(bindingResult.hasErrors()).willReturn(true);

		Model expectedModel = new ExtendedModelMap();
		expectedModel.addAttribute("form", createQuestionForm);
		expectedModel.addAttribute("showBindErrors", true);

		// When
		ModelAndView modelAndView = questionController.createNewQuestion(createQuestionForm, bindingResult, principal);

		// Then
		assertThat(modelAndView)
				.as("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(asList(
						"question/create",
						expectedModel
				));
	}

	@Test
	public void shouldCreateNewQuestionGivenSaveAction() {
		// Given
		BindingResult bindingResult = mock(BindingResult.class);
		Principal principal = mock(Principal.class);
		CreateQuestion createQuestionForm = aSimpleCreateQuestionAboutTriangles()
				.action(CreateQuestion.Action.save)
				.build();

		String username = "aUser";
		given(principal.getName()).willReturn(username);
		given(bindingResult.hasErrors()).willReturn(false);

		Question newQuestion = aSimpleQuestionAboutTriangles()
				.id(null)
				.slug(null)
				.createdBy(username)
				.updatedBy(null)
				.updatedDateTime(null)
				.status(QuestionStatus.SUBMITTED_FOR_APPROVAL)
				.votes(0)
				.build();

		Model expectedModel = new ExtendedModelMap();

		// When
		ModelAndView modelAndView = questionController.createNewQuestion(createQuestionForm, bindingResult, principal);

		// Then
		then(questionService).should().saveNewQuestion(argThat(q -> {
			assertThat(q).isEqualToIgnoringGivenFields(newQuestion, "createdDateTime");
			return true;
		}));
		assertThat(modelAndView)
				.as("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(asList(
						"redirect:/",
						expectedModel
				));
	}

	private CreateQuestion.CreateQuestionBuilder createQuestion() {
		return CreateQuestion.builder()
				.answerOptions(asList(
						generateEmptyAnswerOption(),
						generateEmptyAnswerOption()
				))
				.tags(singletonList(""))
				.certifications(singletonList(""))
				.furtherReadings(singletonList(generateEmptyFurtherReading()));
	}
}

File: exam-slammr-webapp/src/test/kotlin/uk/co/claritysoftware/exam/slammr/webapp/web/controller/QuestionControllerTest.kt
package uk.co.claritysoftware.exam.slammr.webapp.web.controller

import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argThat
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.BDDMockito.given
import org.mockito.BDDMockito.then
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.mock
import org.mockito.junit.MockitoJUnitRunner
import org.springframework.ui.ExtendedModelMap
import org.springframework.validation.BindingResult
import uk.co.claritysoftware.exam.slammr.webapp.service.QuestionService
import uk.co.claritysoftware.exam.slammr.webapp.service.model.question.QuestionStatus.SUBMITTED_FOR_APPROVAL
import uk.co.claritysoftware.exam.slammr.webapp.testsupport.service.model.question.QuestionTestDataFactory.Companion.aSimpleQuestionAboutTriangles
import uk.co.claritysoftware.exam.slammr.webapp.testsupport.web.model.question.CreateQuestionTestDataFactory.Companion.aSimpleCreateQuestionAboutTriangles
import uk.co.claritysoftware.exam.slammr.webapp.web.exception.QuestionNotFoundException
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.AnswerOption
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.CreateQuestion
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.FurtherReading
import java.security.Principal
import java.util.*

/**
 * Unit test class for [QuestionController]
 */
@RunWith(MockitoJUnitRunner::class)
class QuestionControllerTest {

	@Mock
	private val questionService: QuestionService? = null

	@InjectMocks
	private val questionController: QuestionController? = null

	@Test
	fun shouldGetQuestionPage() {
		// Given
		val emptyQuestionForm = CreateQuestion()
		val expectedModel = ExtendedModelMap()
		expectedModel.addAttribute("form", emptyQuestionForm)
		expectedModel.addAttribute("showBindErrors", false)

		// When
		val modelAndView = questionController!!.getQuestionPage()

		// Then
		assertThat(modelAndView)
				.`as`("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(listOf(
						"question/create",
						expectedModel
				))
	}

	@Test
	fun shouldNotCreateNewQuestionGivenAddTagAction() {
		// Given
		val bindingResult = mock(BindingResult::class.java)
		val principal = mock(Principal::class.java)
		val createQuestionForm = CreateQuestion(
				tags = mutableListOf("tag1"),
				action = CreateQuestion.Action.addTag)

		val returnedCreateQuestionForm = CreateQuestion(
				tags = mutableListOf("tag1", ""),
				action = CreateQuestion.Action.addTag)

		val expectedModel = ExtendedModelMap()
		expectedModel.addAttribute("form", returnedCreateQuestionForm)
		expectedModel.addAttribute("showBindErrors", false)

		// When
		val modelAndView = questionController!!.createNewQuestion(createQuestionForm, bindingResult, principal)

		// Then
		assertThat(modelAndView)
				.`as`("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(listOf(
						"question/create",
						expectedModel
				))
	}

	@Test
	fun shouldNotCreateNewQuestionGivenAddCertificationAction() {
		// Given
		val bindingResult = mock(BindingResult::class.java)
		val principal = mock(Principal::class.java)
		val createQuestionForm = CreateQuestion(
				certifications = mutableListOf("certification1"),
				action = CreateQuestion.Action.addCertification)

		val returnedCreateQuestionForm = CreateQuestion(
				certifications = mutableListOf("certification1", ""),
				action = CreateQuestion.Action.addCertification)
		val expectedModel = ExtendedModelMap()
		expectedModel.addAttribute("form", returnedCreateQuestionForm)
		expectedModel.addAttribute("showBindErrors", false)

		// When
		val modelAndView = questionController!!.createNewQuestion(createQuestionForm, bindingResult, principal)

		// Then
		assertThat(modelAndView)
				.`as`("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(listOf(
						"question/create",
						expectedModel
				))
	}

	@Test
	fun shouldNotCreateNewQuestionGivenAddFurtherReadingAction() {
		// Given
		val bindingResult = mock(BindingResult::class.java)
		val principal = mock(Principal::class.java)
		val createQuestionForm = CreateQuestion(
				furtherReadings = mutableListOf(FurtherReading("Some reference", "http://some.location")),
				action = CreateQuestion.Action.addFurtherReading)

		val returnedCreateQuestionForm = CreateQuestion(
				furtherReadings = mutableListOf(FurtherReading("Some reference", "http://some.location"),
						FurtherReading()),
				action = CreateQuestion.Action.addFurtherReading)
		val expectedModel = ExtendedModelMap()
		expectedModel.addAttribute("form", returnedCreateQuestionForm)
		expectedModel.addAttribute("showBindErrors", false)

		// When
		val modelAndView = questionController!!.createNewQuestion(createQuestionForm, bindingResult, principal)

		// Then
		assertThat(modelAndView)
				.`as`("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(listOf(
						"question/create",
						expectedModel
				))
	}

	@Test
	fun shouldNotCreateNewQuestionGivenAddAnswerAction() {
		// Given
		val bindingResult = mock(BindingResult::class.java)
		val principal = mock(Principal::class.java)
		val createQuestionForm = CreateQuestion(
				answerOptions = mutableListOf(AnswerOption("Some answser", true)),
				action = CreateQuestion.Action.addAnswer)

		val returnedCreateQuestionForm = CreateQuestion(
				answerOptions = mutableListOf(AnswerOption("Some answser", true),
						AnswerOption()),
				action = CreateQuestion.Action.addAnswer)

		val expectedModel = ExtendedModelMap()
		expectedModel.addAttribute("form", returnedCreateQuestionForm)
		expectedModel.addAttribute("showBindErrors", false)

		// When
		val modelAndView = questionController!!.createNewQuestion(createQuestionForm, bindingResult, principal)

		// Then
		assertThat(modelAndView)
				.`as`("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(listOf(
						"question/create",
						expectedModel
				))
	}

	@Test
	fun shouldNotCreateNewQuestionGivenSaveActionAndBindingErrors() {
		// Given
		val bindingResult = mock(BindingResult::class.java)
		val principal = mock(Principal::class.java)
		val createQuestionForm = aSimpleCreateQuestionAboutTriangles()
				.copy(summary = "", action = CreateQuestion.Action.save)

		given(bindingResult.hasErrors()).willReturn(true)

		val expectedModel = ExtendedModelMap()
		expectedModel.addAttribute("form", createQuestionForm)
		expectedModel.addAttribute("showBindErrors", true)

		// When
		val modelAndView = questionController!!.createNewQuestion(createQuestionForm, bindingResult, principal)

		// Then
		assertThat(modelAndView)
				.`as`("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(listOf(
						"question/create",
						expectedModel
				))
	}

	@Test
	fun shouldCreateNewQuestionGivenSaveAction() {
		// Given
		val bindingResult = mock(BindingResult::class.java)
		val principal = mock(Principal::class.java)
		val createQuestionForm = aSimpleCreateQuestionAboutTriangles()
				.copy(action = CreateQuestion.Action.save)

		val username = "aUser"
		given(principal.getName()).willReturn(username)
		given(bindingResult.hasErrors()).willReturn(false)

		val newQuestion = aSimpleQuestionAboutTriangles()
				.copy(id = null, slug = null, createdBy = username, updatedBy = null, updatedDateTime = null, status = SUBMITTED_FOR_APPROVAL, votes = 0)

		val savedQuestion = aSimpleQuestionAboutTriangles()
				.copy(createdBy = username, updatedBy = null, updatedDateTime = null, status = SUBMITTED_FOR_APPROVAL, votes = 0)
		given(questionService!!.saveNewQuestion(any()))
				.willReturn(savedQuestion)

		val expectedModel = ExtendedModelMap()

		// When
		val modelAndView = questionController!!.createNewQuestion(createQuestionForm, bindingResult, principal)

		// Then
		then(questionService).should().saveNewQuestion(argThat {
			assertThat(this).isEqualToIgnoringGivenFields(newQuestion, "createdDateTime")
			true
		})
		assertThat(modelAndView)
				.`as`("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(listOf(
						"redirect:/question/5678/triangle-sides-question",
						expectedModel
				))
	}

	@Test
	fun shouldGetExistingQuestion() {
		// Given
		val questionId = "5678"
		val slug = "triangle-sides-question"

		val question = aSimpleQuestionAboutTriangles()
		given(questionService!!.getQuestionById(anyString()))
				.willReturn(Optional.of(question))

		val expectedModel = ExtendedModelMap()
		expectedModel.addAttribute("question", question)

		// When
		val modelAndView = questionController!!.getExistingQuestionPage(questionId, slug)

		// Then
		then(questionService).should().getQuestionById(questionId)
		assertThat(modelAndView)
				.`as`("Expected ModelAndView is returned")
				.extracting("viewName", "model")
				.isEqualTo(listOf(
						"question/view",
						expectedModel
				))
	}

	@Test
	fun shouldNotGetExistingQuestionGivenNonExistentQuestionId() {
		// Given
		val questionId = "1234"
		val slug = "triangle-sides-question"

		given(questionService!!.getQuestionById(anyString()))
				.willReturn(Optional.empty())

		// When
		val throwable = catchThrowable { questionController!!.getExistingQuestionPage(questionId, slug) }

		// Then
		then(questionService).should().getQuestionById(questionId)
		assertThat(throwable)
				.isInstanceOf(QuestionNotFoundException::class.java)
				.hasMessage("Question with id 1234 not found")
	}
}